students/john_navitsky/session02/series.py

```diff
@@ -0,0 +1,49 @@
+
+""" Functions that return Nth entry of a sum sequence. """
+
+def sum_series(nth=1,sequence=[0,1]):
```

Comment:
Careful here!

This will work if you call it once, but can give very wrong results called additional times. Try:

assert sum_series(5) == 3
assert sum_series(6) == 5

It fails! To see what's going on, print the sequence:

```
In [17]: def sum_series(nth=1,sequence=[0,1]):
    ...:     """ Generate a list of sums given a seed and return the Nth number.
    ...:  """
    ...:     for i in range(2, nth):
    ...:         sequence.append(sequence[i-2]+sequence[i-1])
    ...:     print(sequence)
    ...:     return sequence[nth-1]
    ...: 

In [18]: sum_series(5)
[0, 1, 1, 2, 3]
Out[18]: 3

In [19]: sum_series(6)
[0, 1, 1, 2, 3, 1, 2, 3, 5]
Out[19]: 1
```
Why???

We'll get into this in class later, but that default argument: `sequence=[0,1]` gets evaluated ONCE, when the function is defined. so every time the function is used, it re-uses the same list, and keeps adding to it!

As a rule, you don't want to use a mutable object as a default argument. Rather, something like this:

```
def sum_series(nth=1, sequence=None):
    if sequence is None:
        sequence=[0,1]
```

that will guarantee you get a new one each time the function is run.

Another, maybe better option is to use an immutable sequence:

```
def sum_series(nth=1, sequence=(0, 1)):
    sequence = list(sequence)
```
That way, the user can pass in any sequence (tuple or list), and if they pass in a list it won't be mutated by the function.